src/vsr/client.zig

```diff
@@ -26,10 +28,83 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
                 operation: StateMachine.Operation,
                 results: []const u8,
             ) void;
-            user_data: u128,
-            // Null iff operation=register.
-            callback: ?Callback,
+
+            const Info = struct {
+                user_data: u128,
+                callback: ?Callback, // Null iff operation=register.
+            };
+
+            const Demux = struct {
+                next: ?*Demux = null,
+                info: Info,
+                event_count: u16,
+                event_offset: u16,
+            };
+
+            info: Info,
             message: *Message,
+            demux_queue: FIFO(Demux) = .{ .name = null },
+        };
+
+        pub const Batch = struct {
+            request: *Request,
+            demux: ?*Request.Demux, // Null iff batch_get() newly created `request`
+
+            pub fn slice(batch: Batch) []u8 {
+                const body_offset = if (batch.demux) |demux| blk: {
+                    const operation = batch.request.message.header.operation.cast(StateMachine);
+                    switch (operation) {
+                        inline else => |op| {
+                            const event_size: u32 = @sizeOf(StateMachine.Event(op));
+                            break :blk event_size * demux.event_offset;
+                        },
+                    }
+                } else blk: {
+                    assert(batch.request.demux_queue.empty());
+                    break :blk 0;
+                };
+
+                const body = batch.request.message.body();
+                return body[body_offset..];
+            }
+        };
+
+        const DemuxPool = struct {
+            free: ?*Request.Demux = null,
+            used: usize = 0,
+            memory: []Request.Demux,
+
+            fn init(allocator: mem.Allocator) !DemuxPool {
+                return DemuxPool{
+                    .memory = try allocator.alloc(
+                        Request.Demux,
+                        StateMachine.constants.batch_logical_max,
+                    ),
+                };
+            }
+
+            fn deinit(pool: *DemuxPool, allocator: mem.Allocator) void {
+                allocator.free(pool.memory);
+            }
+
+            fn acquire(pool: *DemuxPool) ?*Request.Demux {
+                if (pool.free) |demux| {
+                    pool.free = demux.next;
+                    return demux;
+                }
+                if (pool.used < pool.memory.len) {
+                    defer pool.used += 1;
+                    return &pool.memory[pool.used];
+                }
+                return null;
+            }
```

Comment:
Why use the "lazy fifo" pattern rather than an IOPS? :thinking:

Reply:
IOPS is limited to 2^6 elements 🙁

Reply:
Hmm, is that arbitrary? We could just bump that capacity `u6` up to a large enough int to fix whatever the expected capacity here is. Admittedly it introduces a bitset (which the current `DemuxPool` avoids), but it would also remove ~30 lines of code.

Reply:
IOPS uses an array for the elements and `batch_logical_max` was so large (12MB worth of Demux nodes) that it would make the vsr.Client overflow the stack.